token-limiter-concurrent/src/Control/Concurrent/TokenLimiter/Concurrent.hs

```diff
@@ -133,13 +147,13 @@ waitDebitObserve TokenLimiter {..} debit = modifyMVar tokenLimiterLastServiced $
       let microsecondsToWaitDouble :: Double
           microsecondsToWaitDouble =
             1_000_000
-              -- fromIntegral :: Word64 -> Double
-              * fromIntegral extraTokensNeeded
-              -- fromIntegral :: Word64 -> Double
-              / fromIntegral (tokenLimitConfigTokensPerSecond tokenLimiterConfig)
+              * (fromIntegral :: Word64 -> Double) extraTokensNeeded
+              / (fromIntegral :: Word64 -> Double) (tokenLimitConfigTokensPerSecond tokenLimiterConfig)
+
       let microsecondsToWait = ceiling microsecondsToWaitDouble
-      -- threadDelay guarantees that _at least_ the given number of microseconds will have passed.
-      threadDelay microsecondsToWait
+      withMVar tokenLimiterWaitLock $ \() ->
```

Comment:
@robinp That significantly changes the semantics.
We want the `WaitLock` to only be taken when actual waiting is occurring, otherwise the results of the other functions (that return false if this lock is taken) are different.

Reply:
Well, once the `waitDebit` function is called, there _is_ some actual wait that would be occurring (given a concurrent caller), at most at the limit it is zero-ish (if the current token count suffices). So I don't think it is meaningful to distinguish the zeroish-wait case from the nonzeroish-wait case, from the perspective of concurrent debitors.

So as long as tryTakeMVar is used for canDebit / tryDebit (and the semantic change of "False" as described in other comment is accepted), it should be ok to take the lock here at the beginning.

But I think we are designing a bit in a vacuum - it is hard to tell what is reasonable behavior from the canDebit / tryDebit functions without knowing how the callers would use the result. For example, two main scenarios I can imagine (but I'm not very imaginative now):

A) Caller would anyway need to perform the action after the rate limit. Then not much use in canDebit/tryDebit, just grit teeth and do a waitDebit.

B) Caller would shed actions that can't be immediately actioned.

Maybe the return values of canDebit/tryDebit shouldn't be Bool, but a tri-state of `(Possible)Success|NotEnoughTokens|OtherActionIsWaiting` (where the currently proposed change's `False` just conflates the two latter), and let the caller decide up to its taste how to proceed.

(Btw, something like a `tryModifyMVar` helper that could be written could be utilized to provide this separation)

Reply:
@robinp This has been in my head all evening now.
You've just given me the most useful insight.

We don't need `canDebit` at all, I think it was a misfeature.

I'll tell you the exact behaviour I want.
I have a website (https://social-dance.today) that needs to do geocoding.
There are two geocoding providers:

1. OpenStreetMaps' Nominatim, which is free, but hard-rate-limited to 1 request per second.
2. Google Maps, which is paying.

I have two use cases: Geocoding ASAP (for web requests) and Geocoding whenever (for batch processes).

In the ASAP version, I want to use `tryDebit` to use OpenStreetMaps if the rate limit allows that, and google maps otherwise, to save money.
In the "whenever" version, I want to use `waitDebit` to always OpenStreetMaps and save time that way.

Reply:
If we define "could have debited" as False if the MVar is currently taken, then the `tryModifyMVar` route for `tryDebit` is sufficient.

If that doesn't fly, then sounds like one needs to introduce a "max tolerated timeout" for `tryDebit` to make its decision (and then either enforce it by some combination of takeMVar + Async.race), but it sounds like premature optimization to avoid the small race window of MVar contention in the enough-tokens case.

Probably the API cost you would have to pay over lifetime for those contention-originated calls is less than the code review time cost :)

Reply:
@robinp Where do you see `tryModifyMVar` I can't seem to find it?

Reply:
Aha, was able to write it based on another codebase:

```                                       
tryModifyMVar :: MVar a -> (a -> IO (a, b)) -> IO (Maybe b)               
tryModifyMVar m io =                               
  uninterruptibleMask $ \restore -> do   
    maybea <- tryTakeMVar m                                                  
    forM maybea $ \a -> do                               
      (a', b) <-                                                    
        restore (io a)                                                                                    
          `onException` putMVar m a                                        
      putMVar m a'                       
      return b
```